### PR TITLE
Add DNS Provider for IBM Cloud (SoftLayer)

### LINKED ISSUE #1486: Support for provider: IBM Cloud (SoftLayer)
### Welcome

- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [x] Yes, the DNS provider exposes a public API.
- [X] Yes, I know that the lego maintainers don't have an account in all DNS providers in the world.
- [X] Yes, I'm able to create a pull request and be able to maintain the implementation.
- [X] Yes, I'm able to test an implementation if someone creates a pull request to add the support of this DNS provider.

### How do you use lego?

Docker image

### Link to the DNS provider

https://cloud.ibm.com/docs/dns

### Link to the API documentation

https://cloud.ibm.com/docs/dns?topic=dns-getting-started-with-the-dns-api

### Additional Notes

I'm sorry I'm not good at English

Implementation is over.

I would like to make the first PR in my life
Can you guide me?

### PR DESCRIPTION
**Pull request by ncs-nozomi-nishinohara**
Add DNS Provider for IBM Cloud (SoftLayer)

Close #1486

### CODE REVIEW

**Comment by ncs-nozomi-nishinohara** [comment]
Generate with wildcard (substituted the domain I tested with for example.com):

```
export SOFTLAYER_USERNAME=xxxxxxx
export SOFTLAYER_API_KEY=xxxxxxx

./lego -m nozomi_nishinohara@n-creativesystem.com --dns softlayer -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
2021/09/11 15:04:43 [INFO] [*.example.com, example.com] acme: Obtaining bundled SAN certificate
2021/09/11 15:04:44 [INFO] [*.example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/468256618
2021/09/11 15:04:44 [INFO] [example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/468271928
2021/09/11 15:04:44 [INFO] [example.com] acme: authorization already valid; skipping challenge
2021/09/11 15:04:44 [INFO] [*.example.com] acme: use dns-01 solver
2021/09/11 15:04:44 [INFO] [*.example.com] acme: Preparing to solve DNS-01
2021/09/11 15:04:46 [INFO] [*.example.com] acme: Trying to solve DNS-01
2021/09/11 15:04:46 [INFO] [*.example.com] acme: Checking DNS record propagation using [127.0.0.11:53]
2021/09/11 15:04:48 [INFO] Wait for propagation [timeout: 1m0s, interval: 2s]
2021/09/11 15:04:49 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:04:51 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:04:53 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:04:55 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:04:57 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:04:59 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:01 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:03 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:05 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:07 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:10 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:12 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:14 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:16 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:18 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:20 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:22 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:24 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:26 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:29 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:31 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:33 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:35 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 15:05:44 [INFO] [*.example.com] The server validated our request
2021/09/11 15:05:44 [INFO] [*.example.com] acme: Cleaning DNS-01 challenge
2021/09/11 15:05:47 [INFO] [*.example.com, example.com] acme: Validations succeeded; requesting certificates
2021/09/11 15:05:48 [INFO] [*.example.com] Server responded with a certificate.
```

**Comment by ldez** [comment]
Hello, in order for a PR adding a DNS provider to be accepted, you have to:

- [ ] add a description to your PR
- [ ] be able to maintain this provider
- [ ] have a homogeneous design with the other providers
- [ ] add tests (units)
```shell
make test
```
- [ ] add tests ("live")
https://github.com/go-acme/lego/blob/c938de68f262c057542d44bcb67ef961cefcc60e/providers/dns/cloudflare/cloudflare_test.go#L125-L151

```shell
make test
```
- [ ] add a [provider descriptor](https://github.com/go-acme/lego/blob/master/providers/dns/cloudflare/cloudflare.toml)
- [ ] generate CLI help, documentation, and readme.
```bash
make generate-dns
```
- [ ] be able to do: _(and put the output of this command to a comment in your PR)_
```bash
rm -rf .lego

./lego -m your@email.com --dns YOUR_PROVIDER_NAME -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```
Note the wildcard domain is important.
- [ ] pass the linter ([golangci-lint](https://github.com/golangci/golangci-lint#install) must be installed):
```shell
make checks
```
- [ ] do `go mod tidy`

**Comment by ncs-nozomi-nishinohara** [comment]
@ldez 

> - [ ] be able to maintain this provider
> - [ ] have a homogeneous design with the other providers

What should I do with this?

> - [ ] be able to do: (and put the output of this command to a comment in your PR)

I have posted the result of changing the domain part to example.com. Is this not good?
https://github.com/go-acme/lego/pull/1487#issuecomment-917349597

**Comment by ldez** [comment]
> I have posted the result of changing the domain part to example.com. Is this not good?

You have to do all the steps that are described:
```bash
rm -rf .lego

./lego -m nozomi_nishinohara@n-creativesystem.com --dns softlayer -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```

**Comment by ncs-nozomi-nishinohara** [comment]
@ldez 

```
rm -rf .lego

./lego -m nozomi_nishinohara@n-creativesystem.com --dns softlayer -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```
I ran it.
I don't have a domain with my IBM Cloud account, so I get an error. Is that okay?

```
2021/09/11 22:13:56 No key found for account nozomi_nishinohara@n-creativesystem.com. Generating a P256 key.
2021/09/11 22:13:56 Saved key to /src/dist/.lego/accounts/acme-staging-v02.api.letsencrypt.org/nozomi_nishinohara@n-creativesystem.com/keys/nozomi_nishinohara@n-creativesystem.com.key
2021/09/11 22:13:57 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
Do you accept the TOS? Y/n
y
2021/09/11 22:13:59 [INFO] acme: Registering account for nozomi_nishinohara@n-creativesystem.com
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/src/dist/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2021/09/11 22:13:59 [INFO] [*.example.com, example.com] acme: Obtaining bundled SAN certificate
2021/09/11 22:14:00 [INFO] [*.example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/470171118
2021/09/11 22:14:00 [INFO] [example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/470171128
2021/09/11 22:14:00 [INFO] [*.example.com] acme: use dns-01 solver
2021/09/11 22:14:00 [INFO] [example.com] acme: Could not find solver for: tls-alpn-01
2021/09/11 22:14:00 [INFO] [example.com] acme: Could not find solver for: http-01
2021/09/11 22:14:00 [INFO] [example.com] acme: use dns-01 solver
2021/09/11 22:14:00 [INFO] [*.example.com] acme: Preparing to solve DNS-01
2021/09/11 22:14:01 [INFO] [example.com] acme: Preparing to solve DNS-01
2021/09/11 22:14:02 [INFO] [*.example.com] acme: Cleaning DNS-01 challenge
2021/09/11 22:14:03 [WARN] [*.example.com] acme: cleaning up failed: softlayer: no data found of domain: example.com 
2021/09/11 22:14:03 [INFO] [example.com] acme: Cleaning DNS-01 challenge
2021/09/11 22:14:04 [WARN] [example.com] acme: cleaning up failed: softlayer: no data found of domain: example.com 
2021/09/11 22:14:04 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/470171118
2021/09/11 22:14:05 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/470171128
2021/09/11 22:14:06 Could not obtain certificates:
        error: one or more domains had a problem:
[*.example.com] [*.example.com] acme: error presenting token: softlayer: no data found of domain: example.com
[example.com] [example.com] acme: error presenting token: softlayer: no data found of domain: example.com
```

**Comment by ldez** [comment]
> I don't have a domain with my IBM Cloud account

I'm not sure to understand your sentence.
Do you mean that you don't have a domain or you don't have the domain `example.com`?

`example.com` must be replaced by your real domain.

I updated your PR, could you fetch the new commits and retry?

Note: I change the name of the provider from `softlayer` to `ibmcloud`

```
rm -rf .lego

./lego -m nozomi_nishinohara@n-creativesystem.com --dns ibmcloud -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```

**Comment by ncs-nozomi-nishinohara** [comment]
@ldez 

The actual domain cannot be published because it is my customer's domain.

https://github.com/go-acme/lego/pull/1487#issuecomment-917349597
The result of this is changing the actual domain part to example.com,
Is it always useless if it is not the actual domain?

**Comment by ldez** [comment]
It's not a problem to not give the real domain in the log for this PR.
You are using `example.com` to anonymize your domain in the log is right.

I guess that you are using a real domain to execute the command.

But I need the output of the command for a domain and a wildcard domain at the same time.
You can use a subdomain if you want, ex:
```bash
export SOFTLAYER_USERNAME=xxxxxxx
export SOFTLAYER_API_KEY=xxxxxxx

rm -rf .lego

./lego -m nozomi_nishinohara@n-creativesystem.com --dns ibmcloud -d *.test.example.com -d test.example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```

But I need the output of one command with 2 domains, and one of these domains must be a wildcard domain, and 2 domains must have the same root domain. ex:
- `*.example.com` and `example.com`: OK
- `*.example.com` and `google.org`: NOT OK
- `*.test.example.com` and `test.example.com`: OK
- `*.test.example.com` : NOT OK
- `test.example.com`: NOT OK
- `*.test.example.com` and `test.google.com`: NOT OK

**Comment by ncs-nozomi-nishinohara** [comment]
excuse me.

Is this fine?
I ran it with a real domain wildcard and a real domain.

Or does a domain without wildcards specify a different domain?
```
rm -rf .lego

./lego -m nozomi_nishinohara@n-creativesystem.com --dns ibmcloud -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run

2021/09/11 22:38:55 [INFO] [*.example.com, example.com] acme: Obtaining bundled SAN certificate
2021/09/11 22:38:57 [INFO] [*.example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/470269678
2021/09/11 22:38:57 [INFO] [example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/470277218
2021/09/11 22:38:57 [INFO] [example.com] acme: authorization already valid; skipping challenge
2021/09/11 22:38:57 [INFO] [*.example.com] acme: use dns-01 solver
2021/09/11 22:38:57 [INFO] [*.example.com] acme: Preparing to solve DNS-01
2021/09/11 22:38:58 [INFO] [*.example.com] acme: Trying to solve DNS-01
2021/09/11 22:38:58 [INFO] [*.example.com] acme: Checking DNS record propagation using [127.0.0.11:53]
2021/09/11 22:39:00 [INFO] Wait for propagation [timeout: 1m0s, interval: 2s]
2021/09/11 22:39:01 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:03 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:05 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:07 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:09 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:11 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:13 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:15 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:17 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:19 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:21 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:24 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:26 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:28 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:30 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:32 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:34 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:36 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:38 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:39:41 [INFO] [*.example.com] The server validated our request
2021/09/11 22:39:41 [INFO] [*.example.com] acme: Cleaning DNS-01 challenge
2021/09/11 22:39:44 [INFO] [*.example.com, example.com] acme: Validations succeeded; requesting certificates
2021/09/11 22:39:46 [INFO] [*.example.com] Server responded with a certificate.
```

**Comment by ldez** [comment]
> Or does a domain without wildcards specify a different domain?

Sorry, I don't understand your question, could you rephrase your question?

**Comment by ldez** [comment]
I need the full log (with an anonymized domain).
Because I need to be able to check that you are following all the steps.

**Comment by ncs-nozomi-nishinohara** [comment]
Google Translate changes meaning.

Is this?
The real domain has changed to example.com.
```
root@da1a6db97467:/src/dist# rm -rf .lego
root@da1a6db97467:/src/dist# ./lego -m nozomi_nishinohara@n-creativesystem.com --dns ibmcloud -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
2021/09/11 22:53:41 No key found for account nozomi_nishinohara@n-creativesystem.com. Generating a P256 key.
2021/09/11 22:53:41 Saved key to /src/dist/.lego/accounts/acme-staging-v02.api.letsencrypt.org/nozomi_nishinohara@n-creativesystem.com/keys/nozomi_nishinohara@n-creativesystem.com.key
2021/09/11 22:53:42 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
Do you accept the TOS? Y/n
y
2021/09/11 22:53:43 [INFO] acme: Registering account for nozomi_nishinohara@n-creativesystem.com
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/src/dist/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2021/09/11 22:53:44 [INFO] [*.example.com, example.com] acme: Obtaining bundled SAN certificate
2021/09/11 22:53:45 [INFO] [*.example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/470327108
2021/09/11 22:53:45 [INFO] [example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/470327118
2021/09/11 22:53:45 [INFO] [*.example.com] acme: use dns-01 solver
2021/09/11 22:53:45 [INFO] [example.com] acme: Could not find solver for: tls-alpn-01
2021/09/11 22:53:45 [INFO] [example.com] acme: Could not find solver for: http-01
2021/09/11 22:53:45 [INFO] [example.com] acme: use dns-01 solver
2021/09/11 22:53:45 [INFO] [*.example.com] acme: Preparing to solve DNS-01
2021/09/11 22:53:47 [INFO] [example.com] acme: Preparing to solve DNS-01
2021/09/11 22:53:49 [INFO] [*.example.com] acme: Trying to solve DNS-01
2021/09/11 22:53:49 [INFO] [*.example.com] acme: Checking DNS record propagation using [127.0.0.11:53]
2021/09/11 22:53:51 [INFO] Wait for propagation [timeout: 1m0s, interval: 2s]
2021/09/11 22:53:51 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:53:53 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:53:56 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:53:58 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:00 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:02 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:14 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:16 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:18 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:20 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:22 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:24 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:26 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:29 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/09/11 22:54:40 [INFO] [*.example.com] The server validated our request
2021/09/11 22:54:40 [INFO] [example.com] acme: Trying to solve DNS-01
2021/09/11 22:54:40 [INFO] [example.com] acme: Checking DNS record propagation using [127.0.0.11:53]
2021/09/11 22:54:42 [INFO] Wait for propagation [timeout: 1m0s, interval: 2s]
2021/09/11 22:54:42 [INFO] [example.com] The server validated our request
2021/09/11 22:54:42 [INFO] [*.example.com] acme: Cleaning DNS-01 challenge
2021/09/11 22:54:45 [INFO] [example.com] acme: Cleaning DNS-01 challenge
2021/09/11 22:54:48 [INFO] [*.example.com, example.com] acme: Validations succeeded; requesting certificates
2021/09/11 22:54:49 [INFO] [*.example.com] Server responded with a certificate.
```

**Comment by ldez** [comment]
Thank you, everything seems good now :+1:

**Comment by ncs-nozomi-nishinohara** [comment]
@ldez 

thank you very much!
It was my first contribution to OSS, but I had a good experience!

**Comment by ldez** [comment]
I am glad to hear that your first OSS contribution was a good experience, and I hope you will now contribute to many OSS projects :smile_cat:

**Review by ldez — APPROVED** [review]
